taskcat/taskcat-v9
# Enable CLI and Wire up (linting, lambda_builds, s3staging, validate, stack creation, display_progress, stack deletion)

## Overview
### Items Ready
- Wire up cli
- Enable Linting
- Enable S3 Staging
- Enable Stacks Creation
- Enable Stack Progress Reporting (Stream to Terminal)
- Enable Stack Deletion

### Item to do
**Add Unit Tests as needed**

#### Completed:

###  Expected Output
```

 _            _             _              ___
| |_ __ _ ___| | _____ __ _| |_    __   __/ _ \
| __/ _` / __| |/ / __/ _` | __|___\ \ / / (_) |
| || (_| \__ \   < (_| (_| | ||_____\ V / \__, |
 \__\__,_|___/_|\_\___\__,_|\__|     \_/    /_/

[INFO   ] : Using local source (development mode)
```
## Enable Linting
```
[INFO   ] : Lint passed for test test1 on template /Users/tonynv/demo/nested-create/templates/test.template.yaml
[INFO   ] : Lint passed for test test1 on template /Users/tonynv/demo/nested-create/templates/test.template_nested.yaml
[INFO   ] : Lint passed for test test2 on template /Users/tonynv/demo/nested-create/templates/test.template_nested.yaml
[INFO   ] : Lint passed for test test2 on template /Users/tonynv/demo/nested-create/templates/test.template.yaml
```
## Enable S3 staging
```
[INFO   ] : Creating bucket in us-east-1 for account
[INFO   ] : Staging Bucket: [taskcat-nested-create-b7c5deb5]
[INFO   ] : Enforcing SigV4 requests for bucket $taskcat-nested-create-b7c5deb5
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/ci/input.json
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/ci/upload.yml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/ci/taskcat.yml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template.yaml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template_inner.yaml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template_nested.yaml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template_middle.yaml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template_simple2.yaml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template_middle3.yaml
[S3: -> ] s3://taskcat-nested-create-b7c5deb5/nested-create/templates/test.template_middle2.yaml
```
## Progress display as console stream

> Repaints in-place till all stack reach `CREATE_COMPLETE` of `CREATE_FAILED`
> Create tree with that shows stack hierarchy 
- Marks `Master Stack` with Ⓜ
- Marks `Nested Stack` with Ⓝ
![v9-stack_progress](https://user-images.githubusercontent.com/5912128/64604000-2e6ce180-d376-11e9-9517-83379d8e89f8.gif)
**Not in terminal, reprint now using normal build-in print function.**

### Notes

Optional. Caveats, Alternatives, Other relevant information.

## Testing Instructions

 How to test this PR Start after checking out this branch (bulleted)

 - `cd taskcat-v9`
 - `python ./bin/taskcat-v9 test run ci/taskcat.yml -p tests/data/nested-create`

 Include test case, and expected output
